### PR TITLE
build: various cmake fixes

### LINKED ISSUE #21672: [build] The nvim executable is linked again during `make install`
### Describe the bug

When building & installing neovim from git HEAD, the nvim executable is linked during `make`, and then linked again during `make install`, this takes a lot of time on not so powerful machines like arm dev boards (due to LTO).

### Steps to reproduce

I'm using GCC 8.5.0 and Binutils 2.39 that comes with the system.

```
git clone --depth 1 https://github.com/neovim/neovim
cd neovim
make -j CMAKE_INSTALL_PREFIX=$PREFIX CMAKE_BUILD_TYPE=Release
make install
```

Here is the log of `make install`

```
cd .deps && \
        /usr/bin/cmake3 -G 'Ninja'   \
         /dev/shm/neovim//cmake.deps
-- CMAKE_BUILD_TYPE=Debug
-- Found GNU Make at /gpfs/home/wenblu/.local/bin/gmake
-- Configuring done
-- Generating done
-- Build files have been written to: /dev/shm/neovim/.deps
mkdir -p build
touch build/.ran-deps-cmake
ninja  -C .deps
ninja: Entering directory `.deps'
[2/11] Performing configure step for 'libuv'
loading initial cache file /dev/shm/neovim/.deps/build/tmp/libuv-cache-Debug.cmake
-- summary of build options:
    Install prefix:  /dev/shm/neovim/.deps/usr
    Target system:   Linux
    Compiler:
      C compiler:    /usr/bin/gcc (GNU)
      CFLAGS:

-- Configuring done
-- Generating done
-- Build files have been written to: /dev/shm/neovim/.deps/build/src/libuv-build
[3/11] Performing build step for 'libuv'
[1/37] Building C object CMakeFiles/uv_a.dir/src/strtok.c.o
[2/37] Building C object CMakeFiles/uv_a.dir/src/unix/linux-syscalls.c.o
[3/37] Building C object CMakeFiles/uv_a.dir/src/strscpy.c.o
[4/37] Building C object CMakeFiles/uv_a.dir/src/version.c.o
[5/37] Building C object CMakeFiles/uv_a.dir/src/idna.c.o
[6/37] Building C object CMakeFiles/uv_a.dir/src/random.c.o
[7/37] Building C object CMakeFiles/uv_a.dir/src/inet.c.o
[8/37] Building C object CMakeFiles/uv_a.dir/src/uv-data-getter-setters.c.o
[9/37] Building C object CMakeFiles/uv_a.dir/src/unix/dl.c.o
[10/37] Building C object CMakeFiles/uv_a.dir/src/fs-poll.c.o
[11/37] Building C object CMakeFiles/uv_a.dir/src/unix/getnameinfo.c.o
[12/37] Building C object CMakeFiles/uv_a.dir/src/unix/random-getrandom.c.o
[13/37] Building C object CMakeFiles/uv_a.dir/src/unix/procfs-exepath.c.o
[14/37] Building C object CMakeFiles/uv_a.dir/src/unix/random-devurandom.c.o
[15/37] Building C object CMakeFiles/uv_a.dir/src/unix/proctitle.c.o
[16/37] Building C object CMakeFiles/uv_a.dir/src/unix/getaddrinfo.c.o
[17/37] Building C object CMakeFiles/uv_a.dir/src/unix/random-sysctl-linux.c.o
[18/37] Building C object CMakeFiles/uv_a.dir/src/unix/poll.c.o
[19/37] Building C object CMakeFiles/uv_a.dir/src/timer.c.o
[20/37] Building C object CMakeFiles/uv_a.dir/src/unix/async.c.o
[21/37] Building C object CMakeFiles/uv_a.dir/src/unix/loop.c.o
[22/37] Building C object CMakeFiles/uv_a.dir/src/unix/loop-watcher.c.o
[23/37] Building C object CMakeFiles/uv_a.dir/src/unix/tty.c.o
[24/37] Building C object CMakeFiles/uv_a.dir/src/threadpool.c.o
[25/37] Building C object CMakeFiles/uv_a.dir/src/unix/pipe.c.o
[26/37] Building C object CMakeFiles/uv_a.dir/src/unix/tcp.c.o
[27/37] Building C object CMakeFiles/uv_a.dir/src/unix/process.c.o
[28/37] Building C object CMakeFiles/uv_a.dir/src/unix/thread.c.o
[29/37] Building C object CMakeFiles/uv_a.dir/src/unix/linux-inotify.c.o
[30/37] Building C object CMakeFiles/uv_a.dir/src/unix/signal.c.o
[31/37] Building C object CMakeFiles/uv_a.dir/src/unix/linux-core.c.o
[32/37] Building C object CMakeFiles/uv_a.dir/src/unix/stream.c.o
[33/37] Building C object CMakeFiles/uv_a.dir/src/unix/udp.c.o
[34/37] Building C object CMakeFiles/uv_a.dir/src/uv-common.c.o
[35/37] Building C object CMakeFiles/uv_a.dir/src/unix/core.c.o
[36/37] Building C object CMakeFiles/uv_a.dir/src/unix/fs.c.o
[37/37] Linking C static library libuv_a.a
[4/11] Performing install step for 'libuv'
[0/1] Install the project...
-- Install configuration: "Debug"
-- Up-to-date: /dev/shm/neovim/.deps/usr/include
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/win.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/version.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/unix.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/tree.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/threadpool.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/sunos.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/posix.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/os390.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/linux.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/errno.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/darwin.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/bsd.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv/aix.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/uv.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/share/doc/libuv/LICENSE
-- Up-to-date: /dev/shm/neovim/.deps/usr/lib/pkgconfig/libuv-static.pc
-- Installing: /dev/shm/neovim/.deps/usr/lib/libuv_a.a
-- Up-to-date: /dev/shm/neovim/.deps/usr/lib/cmake/libuv/libuvConfig.cmake
-- Installing: /dev/shm/neovim/.deps/usr/lib/cmake/libuv/libuvConfig-debug.cmake
[6/11] Performing configure step for 'luv-static'
CMake Warning (dev) at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:438 (message):
  The package name passed to `find_package_handle_standard_args` (LIBUV) does
  not match the name of the calling package (Libuv).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  cmake/Modules/FindLibuv.cmake:33 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:68 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:438 (message):
  The package name passed to `find_package_handle_standard_args` (LUAJIT)
  does not match the name of the calling package (LuaJIT).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  cmake/Modules/FindLuaJIT.cmake:55 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:131 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done
-- Generating done
-- Build files have been written to: /dev/shm/neovim/.deps/build/src/luv-static-build
[7/11] Performing build step for 'luv-static'
[1/2] Building C object CMakeFiles/libluv_a.dir/src/luv.c.o
[2/2] Linking C static library libluv.a
[8/11] Performing install step for 'luv-static'
[0/1] Install the project...
-- Install configuration: "Debug"
-- Installing: /dev/shm/neovim/.deps/usr/lib/libluv.a
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/luv/luv.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/luv/util.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/luv/lhandle.h
-- Up-to-date: /dev/shm/neovim/.deps/usr/include/luv/lreq.h
[11/11] cd /dev/shm/neovim/.deps && /usr/bin/cmake -E touch .third-party
ninja  -C build
ninja: Entering directory `build'
[0/2] Re-checking globbed directories...
[6/36] Generating nvim.pot
[36/36] Generating zh_CN.UTF-8.mo
ninja  -C build install
ninja: Entering directory `build'
[0/2] Re-checking globbed directories...
[1/2] Install the project...
-- Install configuration: "Release"
-- Installing: /dev/shm/opt/share/man/man1/nvim.1
-- Installing: /dev/shm/opt/bin/nvim
-- Installing: /dev/shm/opt/lib64/nvim/
[...]
```

Here is the log of the `make` command before the install:

[make.log](https://github.com/neovim/neovim/files/10363696/make.log)

### Expected behavior

The nvim executable is linked only once.

### Neovim version (nvim -v)

`v0.9.0-dev-f7ad46e`

This also occurs for release `0.8.x`, but not for `0.7.x`.

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Rocky Linux 8.5 AMD64

### Terminal name/version

iTerm2

### $TERM environment variable

xterm-256color

### Installation

Build from repo

### PR DESCRIPTION
**Pull request by dundargoc**
- build: various cmake refactors and simplifications

Add STATUS keyword to message to ensure messages are shown in the
correct order.
Remove DEPS_CXX_COMPILER as we don't rely on C++ for any of our
dependencies.
Simplify how msgpack and luv configure options are constructed.
Rely on the default installation for luv instead of manually passing
configure, build and install commands.
Simplify return code conditional.

- build: remove CMAKE_OSX_ARCHITECTURES_ALT_SEP workaround

CMAKE_OSX_ARCHITECTURES_ALT_SEP was defined as a workaround to prevent
the shell from interpreting `;`, which CMake uses as a list separator.
However, the same thing can be achieved by instead passing
CMAKE_OSX_ARCHITECTURES as a cache variable instead, which is a more
idiomatic way of achieving the same thing.

- build: define CMAKE_BUILD_TYPE before adding it to BUILD_TYPE_STRING

The problem with the current setup is that CMAKE_BUILD_TYPE is defined
after BUILD_TYPE_STRING. BUILD_TYPE_STRING will then be empty on the
first run, meaning that dependencies are built without a build type.
However, since CMAKE_BUILD_TYPE is a cache variable its value will
persist in subsequent runs. On the second run BUILD_TYPE_STRING will
have the correct value, but it's a different value from the ones the
dependencies were built with. This will force some dependencies to be
built again.

Fixes https://github.com/neovim/neovim/issues/21672.

### CODE REVIEW

File: cmake.deps/cmake/BuildMsgpack.cmake [hunk 1]
```diff
@@ -1,23 +1,16 @@
...
     -DCMAKE_GENERATOR_PLATFORM=${CMAKE_GENERATOR_PLATFORM}
-    ${BUILD_TYPE_STRING}
-    # Make sure we use the same generator, otherwise we may
-    # accidentally end up using different MSVC runtimes
-    -DCMAKE_GENERATOR=${CMAKE_GENERATOR})
+    ${BUILD_TYPE_STRING})
+
+if(NOT MSVC)
+  list(APPEND MSGPACK_CMAKE_ARGS
+    -DCMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES_ALT_SEP}
```
Comment (kylo252) [inline cmake.deps/cmake/BuildMsgpack.cmake:12]:
```suggestion
    "-DCMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES_ALT_SEP}"
```

File: cmake.deps/CMakeLists.txt [hunk 2]
```diff
@@ -12,13 +12,13 @@ list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake" "${PROJECT_SOURCE_DI
 include(CheckCCompilerFlag)
 include(Util)
 
+set_default_buildtype()
+
 get_property(isMultiConfig GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
 if(NOT isMultiConfig)
   set(BUILD_TYPE_STRING -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE})
```
Comment (kylo252) [inline cmake.deps/CMakeLists.txt:19]: this should be useful

```suggestion
  set(BUILD_TYPE_STRING -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE})
  message(STATUS "CMAKE_CONFIGURATION_TYPES=${CMAKE_CONFIGURATION_TYPES}")
```

File: cmake.deps/cmake/BuildMsgpack.cmake (hunk 1 above)
Comment (dundargoc) [reply cmake.deps/cmake/BuildMsgpack.cmake:12]: ?

File: cmake.deps/CMakeLists.txt (hunk 2 above)
Comment (dundargoc) [reply cmake.deps/CMakeLists.txt:19]: Wait, why?

File: cmake.deps/CMakeLists.txt [hunk 3]
```diff
@@ -107,10 +107,6 @@ endif()
 
 set(DEPS_C_COMPILER "${CMAKE_C_COMPILER}")
 
-if(CMAKE_OSX_SYSROOT)
-  set(DEPS_C_COMPILER "${DEPS_C_COMPILER} -isysroot${CMAKE_OSX_SYSROOT}")
-endif()
-
```
Comment (clason) [inline cmake.deps/CMakeLists.txt:120]: Nope. That breaks dependency builds.

Comment (dundargoc) [reply cmake.deps/CMakeLists.txt:120]: Sure, just wanna check the errors :)

File: cmake.deps/cmake/BuildMsgpack.cmake (hunk 1 above)
Comment (kylo252) [reply cmake.deps/cmake/BuildMsgpack.cmake:12]: it's just a good habit when dealing with cmake to use quotes

File: cmake.deps/CMakeLists.txt (hunk 2 above)
Comment (kylo252) [reply cmake.deps/CMakeLists.txt:19]: yeah, that's why I suggested to print it inside the condition when it's multi-config

Comment (dundargoc) [reply cmake.deps/CMakeLists.txt:19]: Sorry, I don't get it. The condition is `if(NOT isMultiConfig)`, so it won't be printed when it's multi-config?

File: cmake.deps/cmake/BuildMsgpack.cmake (hunk 1 above)
Comment (dundargoc) [reply cmake.deps/cmake/BuildMsgpack.cmake:12]: I'll take the risk if it means less cluttered cmake code ;)

**Comment by dundargoc** [comment]
I was wondering if anyone with a mac, preferably an M1, can try out this PR and see if it works. I'm pretty confident that the workaround I removed shouldn't be needed, but it can't hurt to be safe.

**Comment by clason** [comment]
@carlocab (who I believe contributed the `CMAKE_OSX_ARCHITECTURES_ALT_SEP` thing to make our universal release workflow work)

**Comment by dundargoc** [comment]
Tested release job: https://github.com/dundargoc/neovim/actions/runs/3968011322/jobs/6800640437

**Comment by lewis6991** [comment]
Works for me.

**Pull request merged**